### PR TITLE
Support for SQLite, MySQL, PostgreSQL and Django 2.2, 3.1 and 3.2

### PR DESCRIPTION
**Pull request by AlexHill**
Hi Camron,

I'm opening a new PR and cancelling #2 - it's long out of date at this point.

I would love to get a new PyPI release out with these updates.

I'm also keen to make a pitch to the django-developers list about getting a NaiveDateTimeField merged into Django core based on the code in this repository. Would you be OK with that?

Cheers,
Alex

### CODE REVIEW

File: naivedatetimefield/__init__.py [hunk 1]
```diff
@@ -112,104 +122,138 @@ def to_python(self, value):
...
 
+class NaiveTimezoneMixin(object):
+    def get_tzname(self):
+        if isinstance(self.output_field, NaiveDateTimeField):
+            if self.tzinfo is not None:
+                warnings.warn(
+                    "tzinfo argument provided when truncating a NaiveDateTimeField. "
+                    "This argument will have no effect."
+                )
+            return "UTC"
```
Comment (camflan) [inline naivedatetimefield/__init__.py]: Is there a constant we can use here instead?

File: tests/models.py [hunk 2]
```diff
@@ -6,13 +6,25 @@
 class NaiveDateTimeTestModel(models.Model):
     aware = models.DateTimeField()
     naive = NaiveDateTimeField()
+    timezone = models.CharField(max_length=100, blank=True, default="utc")
```
Comment (camflan) [inline tests/models.py]: Same, is there a constant we can use? If not, should we define our own and make it "UTC" for consistency?

**Review by camflan — APPROVED** [review]
Lots of great work, thanks a lot @AlexHill!

I think that would be great to propose adding to core!

File: tests/models.py (hunk 2 above)
Comment (AlexHill) [reply tests/models.py]: This one really is a one-off for the test model - I think it's OK for this not to be a constant.

File: naivedatetimefield/__init__.py (hunk 1 above)
Comment (AlexHill) [reply naivedatetimefield/__init__.py]: Yes, good call. This is part of a pretty convoluted process to lead the DB backend to produce the right result. I'll look at making this a constant and commenting more thoroughly.

Comment (AlexHill) [reply naivedatetimefield/__init__.py]: OK, I started looking deeper into why various things work and it led me down a rabbit hole. I noticed that if you use the `TIME_ZONE` setting in your database settings dict things break. So we now support that, and the implementation is overall a lot clearer IMO.

File: naivedatetimefield/__init__.py [hunk 3]
```diff
@@ -124,19 +135,19 @@ def to_python(self, value):
     def get_prep_value(self, value):
         return super(DateTimeField, self).get_prep_value(value)
 
+    def get_db_prep_value(self, value, connection, prepared=False):
+        retval = super().get_db_prep_value(value, connection, prepared)
+        return retval
```
Comment (camflan) [inline naivedatetimefield/__init__.py:140]: nit: can you just return this directly instead of assigning to a var?

**Review by camflan — APPROVED** [review]
Very nice!

File: naivedatetimefield/__init__.py (hunk 1 above)
Comment (camflan) [reply naivedatetimefield/__init__.py]: Whoa, nice find! 👍

**Comment by AlexHill** [comment]
Thanks for the review! 😀 

For the purposes of putting together this proposal to django-developers, can I ask what use-cases led you to develop this library?

**Comment by camflan** [comment]
> Thanks for the review! 😀
> 
> For the purposes of putting together this proposal to django-developers, can I ask what use-cases led you to develop this library?

Sure!

I have a project that is focused on analytics of events at distributed locations. It's important that we can query for these events by the _local_ time of the events across all locations so we can compare how things happen locally.

For example, what is happening at lunchtime - so we want to see all events from 11a to 1p at each location. It's _much_ faster to bake this local timestamp into the row and query across them all vs having to run casts in psql or figure out all the time ranges in django while constructing the query.

So our event base model has both `timestamp` and `local_timestamp` fields and this gives us a big performance boost and flexibility for many different use cases.

**Comment by AlexHill** [comment]
Thanks. I use naive fields in two ways. The first is the one you described. 

The other application is scheduling. We have TV screens in locations across the world, and we want them all to say "HAPPY NEW YEAR" when 2022 ticks over wherever they happen to be. The least confusing way of representing this is to store the naive datetime `2022-01-01 00:00:00` once along with the content, and associate a timezone with each location. Then only when we need to work out what should be shown on each individual screen, and at what instant, do we calculate a timestamp based on those two values.

A lot of complexity fell away when we adopted the naive representation - even around little things like querying the database directly, where Django by default represents all datetimes as UTC.

The alternatives in this case are to either pick a standard reference timezone and handle these datetimes very carefully whenever you want to use or display them, or to duplicate the datetime in your data once for each location.

**Pull request merged**